allennlp/data/iterators/data_iterator.py

```diff
@@ -0,0 +1,49 @@
+from typing import Dict, List, Generator
+
+import numpy
+
+from allennlp.data import Dataset
+from allennlp.common import Params
+
+
+class DataIterator:
+    """
+    An abstract ``DataIterator`` class. ``DataIterators`` must implement __call__, which yields
+    batched examples.
+    """
+    def __call__(self, dataset: Dataset) -> Generator[Dict[str, List[numpy.array]], None, None]:
+        """
+        Returns a generator that yields batches over the given dataset, forever.
+
+        Parameters
+        ----------
+        dataset : ``Dataset``
+        """
+        raise NotImplementedError
+
+    def yield_one_pass(self, dataset: Dataset) -> Generator[Dict[str, List[numpy.array]], None, None]:
```

Comment:
I could go either way here.  I originally had it the way you suggest, with a private `_yield_one_pass` method, then changed it to how it is now, thinking that having fewer parameters to `__call__` was probably better.  But, I can easily change it back.  @DeNeutoy, any thoughts?